emacs: Make gccemacs build on darwin

antifuchs opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
To allow macOS/darwin users to build Emacs with the native-comp feature. These changes are all for problems that on Darwin, result in broken binary installs and failures to resolve the emacs derivation.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

antifuchs committed:
gcc: Fix building libgccjit on darwin, using strip -x

The default `strip` invocation tries to strip global symbols from the
library, and refuses because those are indirect symbol table
references.

antifuchs committed:
emacs: Use stdenv.libc instead of .glibc

The latter doesn't exist on Darwin (and, presumably, other targets as
well). This change allows emacsGcc from emacs-overlay to attempt to
build where previously, the derivation would error out.

antifuchs committed:
gcc: On darwin, adjust IDs of installed .so files also

Turns out that libgccjit gets installed as a .so file, which the gcc
builder.sh didn't change: It only touched .dylib files; that means
that anything linking in libgccjit.so would receive an "Image not
found" error at load time.

With this change, we invoke `install_name_tool` on .so files too,
adjusting their dynamic linker ID, so that they too can be found.

antifuchs commented:
@evanrelf I decided to open a real PR & take this here from #93716. Am currently building `libgccjit` from this branch and I suspect that this time, the emacs configure stage won't error out... but it'll be 2 hours /:

evanrelf commented:
@antifuchs Awesome! And thanks for letting me know 🙂

antifuchs commented:
OK, I finally got this to build properly, and it looks like everything is working on this Darwin installation:
* native GUI
* compilation at runtime
* json via libjason

antifuchs commented:
One more thing that is currently failing:

- [ ] `/nix/store/6s7nbin6khpmc6f6nwvcxl3m3h48c3fn-binutils-wrapper-2.31.1/bin/ld: line 207: /nix/store/7gmqwn05rw84kx06wyqhg7ciyn34j0qa-binutils-2.31.1/bin/ld: No such file or directory` - seems like binutils aren't present, preventing `native-compile` from working properly (it compiles to .elc regardless)

antifuchs commented:
So this looks like the default native-compilation `$PATH` that gets set uses the binutils-wrapper `ld`, which exists, but the *wrapped* `ld` doesn't: there's actually a comment about it [in the darwin binutils wrappers](https://github.com/nixos/nixpkgs/blob/1083b435565b0ea0547f270c42277882a514110f/pkgs/os-specific/darwin/binutils/default.nix#L26). I'm not sure what to do about this, or where this comes from (and why AOT compilation during build works, it produces .eln files then!)

Mic92 commented on pkgs/development/compilers/gcc/builder.sh:
```diff
@@ -251,7 +251,7 @@ postInstall() {
 
     if type "install_name_tool"; then
-        for i in "${!outputLib}"/lib/*.*.dylib; do
+        for i in "${!outputLib}"/lib/*.*.dylib "${!outputLib}"/lib/*.so.[0-9]; do
```
Is it a bug in gcc buildsystem that they install libgccjit as .so file?

Mic92 replied:
Looks like this needs to be fixed here: https://github.com/gcc-mirror/gcc/blob/41d6b10e96a1de98e90a7c0378437c3255814b16/gcc/jit/Make-lang.in#L43

antifuchs replied:
Yep, looks like it. They still make weird linux-soname based assumptions around how the structure works, but I think we could just set this as a make variable (or patch it) to fix. Probably the better way to fix it!

Mic92 replied:
They have it conditionals for windows but not for macOS now: https://github.com/gcc-mirror/gcc/blob/229752afe3156a3990dacaedb94c76846cebf132/gcc/jit/Make-lang.in#L48

Mic92 replied:
Anyway if you don't patch the buildsystem would it not be better to rename it from .so to .dylib?

antifuchs commented:
So as far as I can tell, on Darwin you must pass linker options (especially the ones used to find libgccjit) to `ld` via command-line args, and the `LIBRARY_PATH` env variable does nothing, even with wrapper scripts. That complicates post-install AOT compilation a bit. I filed an enhancement request with the gnu emacs crowd, to let users [set driver options](https://gcc.gnu.org/onlinedocs/jit/topics/contexts.html#c.gcc_jit_context_add_driver_option), but until then I don't think this can work... unless we build & ship a custom wrapper script.

antifuchs commented:
Status update on this:
* I have to fix merge conflicts on this.

* There's a change pending that will add the ability to set linker paths directly via an elisp variable, in https://github.com/emacs-mirror/emacs/tree/scratch/native-comp-gcc-driver-options (I'm building emacs against my changes).

* With that change, I have an outstanding change to make use of that facility, which allows native-comp emacs to run and compile elisp on Darwin.

* I also tried actioning @Mic92's suggestion of patching the libgccjit build process, but could not figure out a way yet to make it properly install the versioned dylib file and a symlink to it... something always clobbers that final symlink. I'd prefer to stick to installing the .so for now (as it works regardless of file name), and fix that build in a follow-up PR.

I'll push changes to make this happen as soon as the scratch branch is on the native-comp feature branch, hopefully the next couple of days!

antifuchs committed:
emacs: Set native-comp library path as linker flags instead

Since Darwin's linker does not understand LIBRARY_PATH, we have to set
the library path as explicit linker flags: This requires a very recent
feature/native-comp emacs revision, but it runs on Darwin and
correctly compiles files at runtime.

antifuchs committed:
emacs: Fix paths to bintools on Darwin

This changes PATH such that the correct linker can be found to
construct .eln files at runtime.

antifuchs commented:
Phew! All done now, I think: With the latest HEAD of `feature/native-comp` on the emacs git repo and this change, I can (after 12h of building the world) run an emacs under Darwin that produces natively-compiled files.

I think this is ready to go in, if you all approve (:

antifuchs commented:
There's two follow-up items that still need a bit of work, but I don't think they stand in the way of getting this change mainlined for now (gccemacs builds & runs, which is better than what it did earlier):

* The Application bundle does not run - you see `emacs: dlopen(/nix/store/m8j1l3d71lc2qbh2ygp1idl47nf8rxz4-emacs-gcc-20200814.0/Applications/Emacs.app/Contents/MacOS/../eln-cache/x86_64-apple-darwin19.6.0-9cab85d51a8656a0/lisp-mode-0189ba85598c041b7504f0a916c04219-5bf5ce8701bca2b361b97967a6f95776.eln, 1): image not found`, indicating that a path is wrong (that file really does not exist under that path). Workaround is to run the `emacs` program, which brings up a GUI emacs as normal, but can't be pinned to the Dock.

* The libgccjit build process needs patching, as pointed out above; we're working around the problem here, but the build result really needs to be .dylib instead of .so.

matthewbauer commented on pkgs/build-support/bintools-wrapper/default.nix:
```diff
@@ -145,7 +145,7 @@ stdenv.mkDerivation {
       fi
 
-    '' + (if !useMacosReexportHack then ''
+    '' + (if !useMacosReexportHack then optionalString (!targetPlatform.isDarwin) ''
```
I don't think this is right. We need to wrap ld in binutils to get the right ldflags.

antifuchs replied:
Hm. I've been noodling about this one a while. [This comment on darwin's binutils](https://github.com/nixos/nixpkgs/blob/4073a8b20b3d65dbde91b746cceb61d189bba366/pkgs/os-specific/darwin/binutils/default.nix#L26) says it though: binutils on darwin doesn't provide `ld`, and shipping an `ld` in binutils-wrapper means anything that has the wrapper on `$PATH` will invoke a wrapper that then blindly tries to invoke an unwrapped binary that doesn't exist: The result is that on Darwin, you get "No such file or directory" invoking `ld` and get very confused.

Ericson2314 replied:
@matthewbauer is right. This wrapper wraps that direction which *does* provide `ld` by getting it from Apple's `cctools`.

antifuchs replied:
I think I didn't express this clearly enough: The binutils wrapper for `ld` on darwin, as installed in the unpatched state, is broken:

``` sh
# Running in nixpkgs with this change backed out:
$ nix-build . -A binutils
these paths will be fetched (7.63 MiB download, 47.08 MiB unpacked):
  /nix/store/4z8l6d7p2l31xfwypg2i7ykjar7r3n0x-gettext-0.21
  /nix/store/6an46yg377ccqs5w9na64xmdcp6f2ic4-coreutils-8.31
  /nix/store/8p0g4jqxq1hwr0417miw5chdwkwxqlxn-libc++-7.1.0
  /nix/store/g0cpzdibxd83lzlcyl16zgwa0f95lp4y-libc++abi-7.1.0
  /nix/store/jh5jyg67b9frbvpc08jziqf5n2mnkx8v-libiconv-osx-10.12.6
  /nix/store/k31hbn227zp0pp21lygbh73p1111da6m-binutils-2.31.1
  /nix/store/lnxj6hqrwvi9q55lzk6c5mgy8aw27zq5-zlib-1.2.11
  /nix/store/nb9givrwj0m1jbjf65x4g4i6y1n7pjf0-Libsystem-osx-10.12.6
  /nix/store/pyw7sdlm0pc8a9pl4wfv6bn40igy6c1n-gmp-6.2.0
  /nix/store/qkw4ib1kapm28avkz902p8vicpglg8rv-expand-response-params
  /nix/store/r5p4a9xaa7hf3pdal0zkhc23m609shqn-bash-4.4-p23
  /nix/store/rkca4cz1caays8yc7nfq5c0b3pg97y06-binutils-wrapper-2.31.1
copying path '/nix/store/nb9givrwj0m1jbjf65x4g4i6y1n7pjf0-Libsystem-osx-10.12.6' from 'https://cache.nixos.org'...
copying path '/nix/store/r5p4a9xaa7hf3pdal0zkhc23m609shqn-bash-4.4-p23' from 'https://cache.nixos.org'...
copying path '/nix/store/qkw4ib1kapm28avkz902p8vicpglg8rv-expand-response-params' from 'https://cache.nixos.org'...
copying path '/nix/store/g0cpzdibxd83lzlcyl16zgwa0f95lp4y-libc++abi-7.1.0' from 'https://cache.nixos.org'...
copying path '/nix/store/jh5jyg67b9frbvpc08jziqf5n2mnkx8v-libiconv-osx-10.12.6' from 'https://cache.nixos.org'...
copying path '/nix/store/8p0g4jqxq1hwr0417miw5chdwkwxqlxn-libc++-7.1.0' from 'https://cache.nixos.org'...
copying path '/nix/store/lnxj6hqrwvi9q55lzk6c5mgy8aw27zq5-zlib-1.2.11' from 'https://cache.nixos.org'...
copying path '/nix/store/4z8l6d7p2l31xfwypg2i7ykjar7r3n0x-gettext-0.21' from 'https://cache.nixos.org'...
copying path '/nix/store/pyw7sdlm0pc8a9pl4wfv6bn40igy6c1n-gmp-6.2.0' from 'https://cache.nixos.org'...
copying path '/nix/store/k31hbn227zp0pp21lygbh73p1111da6m-binutils-2.31.1' from 'https://cache.nixos.org'...
copying path '/nix/store/6an46yg377ccqs5w9na64xmdcp6f2ic4-coreutils-8.31' from 'https://cache.nixos.org'...
copying path '/nix/store/rkca4cz1caays8yc7nfq5c0b3pg97y06-binutils-wrapper-2.31.1' from 'https://cache.nixos.org'...
/nix/store/rkca4cz1caays8yc7nfq5c0b3pg97y06-binutils-wrapper-2.31.1
$ ./result/bin/ld
./result/bin/ld: line 207: /nix/store/k31hbn227zp0pp21lygbh73p1111da6m-binutils-2.31.1/bin/ld: No such file or directory
```

That is, on Darwin it installs a wrapper script that clutters up $PATH but doesn't operate correctly because the thing it wraps doesn't exist.

I'm happy to back this diff out, but I do think binutils as installed on Darwin results in a misleading wrapper, which this change was mean to address.

Ericson2314 replied:
OK yeah that makes sense. `bintutils` is different than `stdenv.cc.bintools`, and the latter what what I was talking about and the the thing actually used on Darwin. We might in thte future make the script general robust to missing stuff, in which case `binutils` wouldn't have broken wrapper but just would be missing the things provided by `cctools`.

Ericson2314 replied:
To be clear, I mean what you observed makes senses. You should still revert this line like Matt said. We'd ned to do something else to fixed wrapped plan GNU binutils anyways, and that's best done in a later commit.

matthewbauer commented on pkgs/development/compilers/gcc/common/pre-configure.nix:
```diff
@@ -49,3 +49,11 @@ lib.optionalString (hostPlatform.isSunOS && hostPlatform.is64bit) ''
+# installation must not remove external symbols, otherwise the install
+# step errors with "symbols referenced by indirect symbol table
+# entries that can't be stripped".
++ lib.optionalString (hostPlatform.isDarwin) ''
```
Can we make this conditional on langJit?

antifuchs replied:
Thanks for the suggestion, updated.

antifuchs committed:
gcc: Only use strip -x if building libgccjit on darwin

It's not necessary to use strip -x otherwise, so let's just use it for
the JIT library.

mjlbach commented on pkgs/applications/editors/emacs/generic.nix:
```diff
@@ -67,18 +67,18 @@ in stdenv.mkDerivation {
+          "${lib.getLib libgccjit}/lib"
+          "${lib.getLib libgccjit}/lib/gcc/${targetPlatform.config}/${libgccjit.version}"
+          "${lib.getLib stdenv.cc.cc}/lib"
+          "${lib.getLib stdenv.libc}/lib"
```
Should this be glibc? 

```
nix-repl> :l <nixpkgs>
[0.0 MiB DL]Added 12314 variables.

nix-repl> lib.getLib stdenv.libc
error: attribute 'libc' missing, at (string):1:12
```

antifuchs replied:
Ah, so that should be a conditional! Good to know. Will correct.

Ericson2314 replied:
`stdenv.cc.libc` is the portable way to do this.

mjlbach commented on pkgs/applications/editors/emacs/generic.nix:
```diff
@@ -67,18 +67,18 @@ in stdenv.mkDerivation {
-        "${lib.getLib stdenv.glibc}/lib"
-      ];
+      libPath = (lib.concatStringsSep " "
+        (builtins.map (x: ''\"-L${x}\"'') [
```
Wondering if we should use -B here

antifuchs replied:
Do you mean for binutils's `ld`? Darwin's `ld` does not support any `-B` options, as far as I can tell.

matthewbauer replied:
I think `-B` is a CC flag, so it's gcc that handles it. I think gccemacs only cares about library files, so I'm not sure if the `-L` vs. `-B` is relevant though.

antifuchs replied:
Ah, it may depend on whether libgccjit recognizes that option as a "driver" option (vs. a "frontend" option I guess?) - the [manual warns](https://gcc.gnu.org/onlinedocs/jit/topics/contexts.html#c.gcc_jit_context_add_driver_option) that only options for the assembler or linker will be meaningful as "driver options". Definitely something to experiment with, though!

mjlbach replied:
Yes, the guix packaging for native-comp explicitly uses the -B option https://github.com/flatwhatson/guix-channel/commit/f9bf0a476cb37a3718dfb9d6a78f22c33504ae97 else it complains about not finding crti.o, crtbeginS.o, and -lgcc. I haven't been able to compile this succesfully else I'd test.

antifuchs replied:
Fascinating! I'll try that - might also help us get rid of the need to use wrapProgram to set paths to the linker and assembler.

mjlbach replied:
Ok, just confirmed that on linux building with the -B option is necessary for straight.el to compile packages outside of nix

antifuchs committed:
emacs: Use stdenv.cc.libc instead of plain .libc

This is the portable way to address the system-appropriate libc,
better than conditionalizing by (darwin/GNU) system or using either
one.

mjlbach commented:
I'm not able to test this branch locally on linux as p11-kit tests fail while building emacs. This isn't happening on master for me.
```
nix-shell -I nixpkgs=https://github.com/antifuchs/nixpkgs/archive/4334b630df36caad93975ad2d62d3b29109b1a6c.tar.gz -p p11-kit
```

```
test-conf: common/test.c:612: p11_test_copy_setgid: Assertion `false && "this code should not be reached"' failed.
./build/litter/tap-driver.sh: line 639: 56529 Aborted                 (core dumped) "$@"
ERROR: test-conf - missing test plan

============================================================================
Testsuite summary for p11-kit 0.23.21
============================================================================
# TOTAL: 743
# PASS:  741
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 2
============================================================================
See ./test-suite.log
Please report to https://github.com/p11-glue/p11-kit/issues
============================================================================
make[4]: *** [Makefile:4928: test-suite.log] Error 1
make[4]: Leaving directory '/build/source'
make[3]: *** [Makefile:5036: check-TESTS] Error 2
make[3]: Leaving directory '/build/source'
make[2]: *** [Makefile:5640: check-am] Error 2
make[2]: Leaving directory '/build/source'
make[1]: *** [Makefile:4813: check-recursive] Error 1
make[1]: Leaving directory '/build/source'
make: *** [Makefile:5643: check] Error 2
builder for '/nix/store/rll2vsqslai9kwn1isq6r70danfkq06z-p11-kit-0.23.21.drv' failed with exit code 2
error: build of '/nix/store/rll2vsqslai9kwn1isq6r70danfkq06z-p11-kit-0.23.21.drv' failed
```

antifuchs commented:
> I'm not able to test this branch locally on linux as p11-kit tests fail while building emacs. This isn't happening on master for me.

That's curious - I could build this derivation just fine on Ubuntu 18.04. I am, however, getting a weird test failure in something else that this PR doesn't touch /:

matthewbauer commented on pkgs/applications/editors/emacs/generic.nix:
```diff
@@ -158,7 +158,7 @@ in stdenv.mkDerivation {
 
     (lib.optionalString nativeComp ''
-      wrapProgram $out/bin/emacs-* --prefix PATH : "${lib.makeBinPath [ binutils binutils-unwrapped ]}"
+      wrapProgram $out/bin/emacs-* --prefix PATH : "${lib.makeBinPath [ clang.bintools binutils binutils-unwrapped ]}"
```
```suggestion
      wrapProgram $out/bin/emacs-* --prefix PATH : "${lib.makeBinPath [ stdenv.cc.bintools stdenv.cc.bintools.bintools ]}"
```

matthewbauer replied:
This avoids pulling in clang when we aren't using it

antifuchs replied:
Awesome, that does look far better - I'll try this when my current build finishes.

antifuchs replied:
Done - I removed the `clang` and `binutils*` args above too, as they were made unnecessary by this change.

antifuchs committed:
emacs: Use stdenv's bintools instead of clang

This way, we don't have to drag clang or binutils/binutils-wrapped
into the emacs closure, and can instead rely on using the correct one
for the platform we're running on.

Co-authored-by: Matthew Bauer <mjbauer95@gmail.com>

antifuchs committed:
emacs: use -B flags for native compilation dependencies

The -B flag to gcc (and libgccjit) allows us to specify where it can
find things it needs to correctly compile code (both programs and
libraries) without adjusting any environmental flags: So, no need to
wrap the program for a PATH entry containing binutils, and no need to
explicitly pass a linker path anymore.

antifuchs committed:
emacs: use -B flags for native compilation dependencies

The -B flag to gcc (and libgccjit) allows us to specify where it can
find things it needs to correctly compile code (both programs and
libraries) without adjusting any environmental flags: So, no need to
wrap the program for a PATH entry containing binutils, and no need to
explicitly pass a linker path anymore.

mjlbach commented on pkgs/applications/editors/emacs/generic.nix:
```diff
@@ -67,18 +67,18 @@ in stdenv.mkDerivation {
+        (builtins.map (x: ''\"-L${x}\"'') [
+          "${lib.getLib libgccjit}/lib"
+          "${lib.getLib libgccjit}/lib/gcc/${targetPlatform.config}/${libgccjit.version}"
+          "${lib.getLib stdenv.cc.cc}/lib"
```
I'm not sure `"${lib.getLib stdenv.cc.cc}/lib"` needs to be included. I believe adisbladis copied these based on discussion in the emacs-overlay repo from a comment I left, and I at the time didn't realize the gcc libraries did not need to be in LIBRARY_PATH. I tested without, and it seems to be fine.

mjlbach commented on pkgs/applications/editors/emacs/generic.nix:
```diff
@@ -67,18 +67,18 @@ in stdenv.mkDerivation {
+        (builtins.map (x: ''\"-L${x}\"'') [
+          "${lib.getLib libgccjit}/lib"
+          "${lib.getLib libgccjit}/lib/gcc/${targetPlatform.config}/${libgccjit.version}"
+          "${lib.getLib stdenv.cc.cc}/lib"
```
I'm not sure `"${lib.getLib stdenv.cc.cc}/lib"` is necessary, I don't know why I had this in LIBRARY_PATH originally. Tested without and seems fine, but also harmless to leave in.

mjlbach replied:
I think -B is supposed to be recursive, previously when passing the library_path as an environmental variable to comp-async-workers (which never worked, but worked if you wrapped emacs) we had to pass subdirectories manually. It's unclear to me if this is still necessary (I lean towards no)

antifuchs replied:
Welp, I did a (not so quick) test run just now, and it seems we do need that specific directory - but we do not need the targetPlatform/version games. The others, on the other hand - can mostly go.

mjlbach replied:
Yes, you seem to be right -- I don't think we need the machine version/target platform based on the manual: 

```
-Bprefix
This option specifies where to find the executables, libraries, include files, and data files of the compiler itself.

The compiler driver program runs one or more of the subprograms cpp, cc1, as and ld. *It tries prefix as a prefix for each program it tries to run, both with and without ‘machine/version/’ for the corresponding target machine and compiler version.*
```

mjlbach replied:
I think we still need three directories: 

```
 "${lib.getLib libgccjit}/lib"
 "${lib.getLib libgccjit}/lib/gcc"
 "${lib.getLib stdenv.cc.libc}/lib"
```

antifuchs replied:
Yup - that's the ones I have now, plus the bin dirs from the platform cc & bintools. I just got that working with an unwrapped emacs binary, so I think that change I just pushed should be it.

rcorrear commented:
Hi all! Been following the discussion lately, haven't tried it but thanks for all the effort! A question comes to mind, has anyone tried the patch for "emacsMac" here?

mjlbach commented:
@rcorrear If you mean, will native-compilation (gccemacs) work with the Mitsuharu's emacs-mac repository, the answer is no. Native compilation is currently in a feature branch, and it would be nontrivial to merge mitsuharu's work. You'll likely have to wait until native compilation is merged into emacs.

rcorrear commented:
@mjlbach yeah, that's what I wanted to know, thanks!

adelbertc commented:
Apologies if this is an inappropriate place to ask, but does anyone know if there is a cached build of emacsGcc on macOS available? I just tried building it with something like:

```nix
    emacsPackagesNg =
      let
        emacs-overlay =
          import (builtins.fetchTarball https://github.com/nix-community/emacs-overlay/archive/master.tar.gz);

        pkgs = import (fetchGit {
          url = "https://github.com/NixOS/nixpkgs.git";
          rev = "6fc3562432357cdbcbb30804b47feacfac452acb"; # this is this PR's merge commit
        }) {
          overlays = [ emacs-overlay ];
        };
      in
        pkgs.emacsPackagesNgGen pkgs.emacsGcc;
```

on a 2017 13" MacBook Pro and it was compiling from like 11:30AM - 7:30PM before I Ctrl+C'd it, and even after all that time it was still building LLVM things 😬

mjlbach commented:
If you want, you can use my overlay which has CI for macOS (only on the linked branch).

https://github.com/mjlbach/emacs-pgtk-nativecomp-overlay/tree/cleanups

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/lsopenurlswithrole-error-opening-nix-installed-application-emacs-native-comp-on-macos/9132/1
